File: killer_sudoku/solver/engine/rules/_labels.py
"""Shared label-formatting helpers for coaching rule hint explanations.

Centralises the four formatting functions that were previously duplicated
in must_contain_outie, cage_confinement, naked_pair, and hidden_single.
"""

from __future__ import annotations

from killer_sudoku.solver.engine.types import Cell, Unit, UnitKind


def unit_label(unit: Unit) -> str:
    """Human-readable label for a unit: 'row 1', 'column 5', 'box 3', 'this cage'."""
    if unit.kind == UnitKind.ROW:
        return f"row {unit.unit_id + 1}"
    if unit.kind == UnitKind.COL:
        return f"column {unit.unit_id - 9 + 1}"
    if unit.kind == UnitKind.BOX:
        return f"box {unit.unit_id - 18 + 1}"
    return "this cage"


def cell_label(cell: Cell) -> str:
    """rNcM notation (1-based) for a cell: e.g. (2, 4) → 'r3c5'."""
    r, c = cell
    return f"r{r + 1}c{c + 1}"


def unit_type_label(kind: UnitKind) -> str:
    """Plural noun for a unit type: 'rows', 'columns', or 'boxes'."""
    if kind == UnitKind.ROW:
        return "rows"
    if kind == UnitKind.COL:
        return "columns"
    return "boxes"


def type_unit_id(kind: UnitKind, r: int, c: int) -> int:
    """Board unit_id of the same-type unit containing cell (r, c).

    ROW  → unit_ids 0–8
    COL  → unit_ids 9–17
    BOX  → unit_ids 18–26
    """
    if kind == UnitKind.ROW:
        return r
    if kind == UnitKind.COL:
        return 9 + c
    return 18 + (r // 3) * 3 + (c // 3)

File: web/src/engine/rules/_labels.ts
/**
 * Label helpers shared by all rule hint implementations.
 *
 */

import type { Cell } from '../types.js';
import { UnitKind } from '../types.js';
import type { Unit } from '../types.js';

export function cellLabel([r, c]: Cell): string {
  return `r${r + 1}c${c + 1}`;
}

const WINDOW_LABELS_BY_CORNER: ReadonlyMap<string, string> = new Map([
  ['1,1', 'top-left window'],
  ['5,1', 'bottom-left window'],
  ['1,5', 'top-right window'],
  ['5,5', 'bottom-right window'],
]);

export function unitLabel(unit: Unit): string {
  const cells = unit.cells as Cell[];
  switch (unit.kind) {
    case UnitKind.ROW: return `row ${cells[0]![0] + 1}`;
    case UnitKind.COL: return `col ${cells[0]![1] + 1}`;
    case UnitKind.BOX: {
      const windowLabel = WINDOW_LABELS_BY_CORNER.get(`${cells[0]![0]},${cells[0]![1]}`);
      if (windowLabel !== undefined) return windowLabel;
      const br = (cells[0]![0] / 3 | 0) + 1;
      const bc = (cells[0]![1] / 3 | 0) + 1;
      return `box (${br},${bc})`;
    }
    default: {
      const labels = cells.sort((a, b) => a[0] - b[0] || a[1] - b[1]).map(cellLabel);
      return `cage [${labels.join(', ')}]`;
    }
  }
}

export function unitTypeLabel(kind: UnitKind): string {
  switch (kind) {
    case UnitKind.ROW: return 'rows';
    case UnitKind.COL: return 'columns';
    case UnitKind.BOX: return 'boxes';
    default: return 'cages';
  }
}

/** Returns the unit ID of a given kind for cell (r, c). */
export function typeUnitId(kind: UnitKind, r: number, c: number): number {
  switch (kind) {
    case UnitKind.ROW: return r;
    case UnitKind.COL: return 9 + c;
    case UnitKind.BOX: return 18 + (r / 3 | 0) * 3 + (c / 3 | 0);
    default: throw new Error(`typeUnitId: unexpected kind ${kind}`);
  }
}
